Dynamic memory snapshot

xbauch opened the pull request:
Enable tracking information about dynamically allocated memory from `memory-analyzer` across to `goto-harness` via memory snapshots. The knowledge about what addresses have been allocated dynamically (and the scope of these allocation in byte size) is collected by breaking the execution at `malloc` call-sites and inspecting the registers of these function calls.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -29,6 +31,66 @@ gdb_value_extractort::gdb_value_extractort(
+bool gdb_value_extractort::memory_scopet::contains(
+  const memory_addresst &point) const
+{
+  size_t begin_int = std::strtoul(begin.address_string.c_str(), NULL, 0);
```
This is a unchecked conversion effectively (`size_t` is not necessarily the same size as `unsigned long`, even though it's often the case). I'd recommend using `safe_string2size_t` from `string2int.h` (same with all the other string to int conversions here).

xbauch replied:
I wasn't sure if the `safe`-conversion would work on hex-numbers.

hannes-steffenhagen-diffblue replied:
They should, they use `stoll` which basically works the same way as the `str*` functions except it's easier to detect failures.

tautschnig replied:
+1

xbauch replied:
Done and refactored so that the conversion is computed once in case of `distance`. See: https://github.com/diffblue/cbmc/pull/4578/commits/3812c28519a207a0e5f247f4b7fe2cd4db19af2b.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -91,6 +91,70 @@ class gdb_value_extractort
 
+  struct memory_scopet
+  {
+    memory_addresst begin;
```
Why not `private`, if these are not arbitrary (as the deleted default constructor implies)?

xbauch replied:
Done, also added getters when necessary.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -91,6 +91,70 @@ class gdb_value_extractort
+  optionalt<mp_integer> get_malloc_size(irep_idt name);
+
+  /// Build the pointee string for address \p point assuming it points to a
+  ///   dynamic allocation of `n' elements each of size \p member_size. E.g.:
```
what's `n` here?

xbauch replied:
`T* point = (T*) malloc(sizeof(T) * n);` .. that `n`.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -230,8 +242,11 @@ code_blockt memory_snapshot_harness_generatort::add_assignments_to_globals(
     [this](const snapshot_pairt &left, const snapshot_pairt &right) {
-      return pointer_depth(left.second.symbol_expr().type()) <
-             pointer_depth(right.second.symbol_expr().type());
+      if(refers_to(right.second.value, left.first))
```
Can you explain what this is for?

xbauch replied:
For example, say `int* p = malloc(8); int *q = p+1;` is the source code for which we want to take the memory snapshot. Then the snapshot will contain: `int tmp[2]; int* p = tmp[0]; int* q = tmp[1];` And both `p` and `q` `refer_to` `tmp`. I sort the symbols based on this notion of _reference_ so that when building the harness entry function, `tmp` is assigned before the two pointers.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/gdb_api.h:
```diff
@@ -81,10 +81,10 @@ class gdb_apit
-    const std::string pointee;
-    const std::string character;
-    const optionalt<std::string> string;
+    memory_addresst address;
```
These fields are all related, so I don't think they should be both be constant and public at the same time.

xbauch replied:
Well they are not `const` in the final version.

hannes-steffenhagen-diffblue commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -433,11 +440,22 @@ exprt gdb_value_extractort::get_non_char_pointer_value(
   if(pointer_value.pointee.find("+") != std::string::npos)
     return true;
 
+  if(pointer_value.pointee.empty())
```
why is this only done if the `pointee` is empty? What does that mean?

xbauch replied:
Well `pointer_value` is the result of the gdb query `get_memory`. When the `pointee` is empty it means that gdb did not know what _structure_ the pointer points to. This can happen if the pointer points to inside a dynamically allocated memory.

hannes-steffenhagen-diffblue reviewed (approved):
Looks ok to me (with the disclaimer that I don't really understand this)

tautschnig commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -91,6 +91,70 @@ class gdb_value_extractort
+    memory_scopet() = delete;
+    memory_scopet(
+      const memory_addresst &begin,
+      mp_integer byte_size,
```
`const mp_integer &byte_size`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -91,6 +91,70 @@ class gdb_value_extractort
+    memory_scopet(
+      const memory_addresst &begin,
+      mp_integer byte_size,
+      irep_idt name)
```
`const irep_idt &name`

xbauch replied:
Done.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -29,6 +31,66 @@ gdb_value_extractort::gdb_value_extractort(
+  CHECK_RETURN(contains(point));
+  size_t begin_int = std::strtoul(begin.address_string.c_str(), NULL, 0);
+  size_t point_int = std::strtoul(point.address_string.c_str(), NULL, 0);
+  return (point_int - begin_int) / member_size;
```
Shouldn't you first do `CHECK_RETURN(point_int >= begin_int);`?

xbauch replied:
That was implied by `contains`. But I refactored this section a bit to avoid duplicate computation of the address-to-string conversion. See: https://github.com/diffblue/cbmc/pull/4578/commits/3812c28519a207a0e5f247f4b7fe2cd4db19af2b.

tautschnig commented on src/memory-analyzer/analyze_symbol.h:
```diff
@@ -91,6 +91,70 @@ class gdb_value_extractort
+    mp_integer byte_size;
+    irep_idt name;
+
+    memory_scopet() = delete;
```
I'm not sure that's necessary if you explicitly define some other constructor?

xbauch replied:
You are right. Removed.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -112,6 +112,22 @@ void gdb_value_extractort::analyze_symbols(
     values.insert({value.address, symbol_expr});
+
+    const symbolt &symbol = ns.lookup(id);
+    const symbol_exprt actual_expr = symbol.symbol_expr();
```
You can delay this until after the below `if`, because you can just use `symbol.type` instead of `actual_expr.type()`.

xbauch replied:
Delayed.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -112,6 +112,22 @@ void gdb_value_extractort::analyze_symbols(
+    const symbol_exprt actual_expr = symbol.symbol_expr();
+    if(actual_expr.type().id() != ID_pointer)
+    {
+      memory_map[id2string(id)] = value;
```
I'm not entirely sure this should be indexed by `std::string`, wouldn't `irep_idt` be preferable?

xbauch replied:
Works with `irep_idt` as well. Changed.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -330,13 +331,18 @@ exprt gdb_value_extractort::get_pointer_to_member_value(
   // temporary object for it: get the type from symbol table, query gdb for
   // value, allocate new object for it and then store into assignments
   if(it == values.end())
+  if(struct_symbol->symbol_expr().type().id() == ID_pointer)
```
This looks very weird?!

xbauch replied:
What about `struct_symbol->type.id()` ? Changed.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -303,6 +303,7 @@ exprt gdb_value_extractort::get_pointer_to_member_value(
   }
 
   if(!found)
+  if(struct_symbol->symbol_expr().type().id() == ID_array)
```
Looks weird!?

xbauch replied:
As above.

tautschnig commented on src/memory-analyzer/analyze_symbol.cpp:
```diff
@@ -291,6 +291,8 @@ exprt gdb_value_extractort::get_pointer_to_member_value(
   CHECK_RETURN(maybe_struct_size.has_value());
   for(const auto &value_pair : values)
+
+  if(memory_map.count(struct_name) == 0)
```
This looks weird.

xbauch replied:
Wrapped in a helper function `has_known_memory_location`.

tautschnig commented on regression/snapshot-harness/static-array-float/test.desc:
```diff
@@ -0,0 +1,13 @@
+FUTURE
```
Please explain (after another `--` at the end of the file) what isn't supported here so far.

xbauch replied:
Added.

tautschnig commented on regression/memory-analyzer/pointer_to_struct_01/test.desc:
```diff
@@ -1,8 +1,7 @@
-tmp = \{ \.next=\(\(struct S \*\)0\) \};
-p = \&tmp;
+st = \{ .next=\&st \};
+p = \&st;
```
Could you please squash this into the commit that made this possible?

xbauch replied:
It was this one: https://github.com/diffblue/cbmc/pull/4578/commits/33d0a79b6843c2c6c1b60ee8c8540dfe9df9da86. I will squash them together.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 41d398d0955cd267bc7fd331184a007cc0a1202c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113173659
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 356276ae695396ce3c7849e4828d7dc1b840a7ce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113173817
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

xbauch commented:
@tautschnig Our initial approach at querying the GDB for malloc info turned out rather un-portable so we re-implemented the collection this info via breaking at malloc call-sites and parsing the registers (starting at this commit: https://github.com/diffblue/cbmc/pull/4578/commits/e6684572765628814daf1a67397b5d085f31c4b2). The current implementation is still lacking in robustness (we do not track calls of `free`, the name `malloc` is hard-coded, etc.) but I wasn't sure how much more we should include in this PR.

Previously, you reviewed it up to https://github.com/diffblue/cbmc/pull/4578/commits/28ad6d67921b6447ea9d87a8721ce4261b4cfb33. I can do more squashing but I first wanted to ask if our present approach is acceptable and if we should make it more robust in this PR.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6bb1d2827c55e6e961e6d94f38a3574c53c7c5e8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113173970

petr-bauch committed:
Fix the ordering in which symbols are initialised

In memory-snapshot-harness: so that pointer to objects are initialised after
those objects.

petr-bauch committed:
Add documentation to gdb api

no functional change

petr-bauch committed:
Un-constify pointer_valuet data

So that we can assume the implicit assignment operator.

petr-bauch committed:
Helpers to extract values from gdb record

One for value, one for register-value.

petr-bauch committed:
Collecting the malloc information directly

by breaking the execution at malloc entry and extracting the byte size and
memory address from registers.

petr-bauch committed:
Reimplement query-malloc-size

by returning the exact value rather than an estimate collected during preceding
execution.

petr-bauch committed:
Introduce facilities for tracking dynamically allocated memory

memory_scopet keeps the starting point and allocated size for a malloc(ed) site.
We also include helper methods to query about dynamic allocation.

petr-bauch committed:
Re-implement points_to_member query

To account for the possibility of the pointee to be dynamically allocated.

petr-bauch committed:
Initialise the databases for dynamic allocation and known symbols

Via dynamically_allocated and memory_map.

petr-bauch committed:
Wrap extracting the type size

into a member method.

petr-bauch committed:
Remove building dummies for pointee objects

Unnecessary now that we enforce analysing those objects.

petr-bauch committed:
Refactor get-pointer-to-member-value

1: If pointee is known, return it's value instead.
2: malloc-size is precise and type-size has a wrapper
3: do not cast values from dynamically allocated pointers: we would loose the
   information about them being arrays.

petr-bauch committed:
Refactor get-pointer-value

1: Use memory map to get pointer-values (but double-check if gdb previously
   claimed that their pointees are themselves)
2: add more checks before casting.

petr-bauch committed:
Make the assignments into a map

instead of a vector to prevent duplications.

petr-bauch committed:
Update older tests

1: Casts are a part of the output
2: We can use some of the assertions previously left out (now that we get the
dynamic allocation right).

petr-bauch committed:
More tests for static and dynamic arrays

Floating point numbers cannot be extracted for now.

tautschnig commented on src/goto-cc/Makefile:
```diff
@@ -57,6 +57,9 @@ endif
 ###############################################################################
 
+goto-gcc$(EXEEXT): goto-cc$(EXEEXT)
+	@ln -sf goto-cc$(EXEEXT) goto-gcc$(EXEEXT)
```
I think make has `$(LN)` to improve portability, and I'd suggest not to hide this output, i.e., please remove the `@` - simply because we don't do quiet output elsewhere either.

tautschnig commented on src/goto-cc/Makefile:
```diff
@@ -48,7 +48,7 @@ include ../common
 all: goto-cl$(EXEEXT)
 endif
-all: goto-cc$(EXEEXT)
+all: goto-gcc$(EXEEXT)
```
Could you please change the lines above so that this is within an `else`/`endif` instead of being done for MSVC as well? Furthermore, please add `goto-gcc$(EXEEXT)` to `CLEANFILES` above.

tautschnig commented:
@xbauch My apologies for taking this long! You might want to squash some commits and then just merge. Thank you for all the work!!!

petr-bauch committed:
Build goto-gcc during compilation

rather than as a part of regression testing and only under non-MSVC environment.

tautschnig commented:
@xbauch Please rebase now that #4729 is in to make CI happy.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ba8beb533cbf239a1eb65fc3e2a8be2389f24a7d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113721650
